lib/Monads/trace/Trace_In_Monad.thy

```diff
@@ -0,0 +1,141 @@
+  by (simp add: whenE_def in_returnOk)
+
+lemma inl_whenE:
+  "((Inl x, s') \<in> mres (whenE P f s)) = (P \<and> (Inl x, s') \<in> mres (f s))"
+  by (auto simp add: in_whenE)
+
+lemma inr_in_unlessE_throwError[termination_simp]:
+  "(Inr (), s') \<in> mres (unlessE P (throwError E) s) = (P \<and> s'=s)"
+  by (simp add: unlessE_def returnOk_def throwError_def in_return)
+
+lemma in_fail:
+  "r \<in> mres (fail s) = False"
+  by (simp add: fail_def mres_def)
+
+lemma in_assert:
+  "(r, s') \<in> mres (assert P s) = (P \<and> s' = s)"
+  by (simp add: assert_def return_def fail_def mres_def)
+
+lemma in_assertE:
+  "(r, s') \<in> mres (assertE P s) = (P \<and> r = Inr () \<and> s' = s)"
+  by (simp add: assertE_def returnOk_def return_def fail_def mres_def)
+
+lemma in_assert_opt:
+  "(r, s') \<in> mres (assert_opt v s) = (v = Some r \<and> s' = s)"
+  by (auto simp: assert_opt_def in_fail in_return split: option.splits)
+
+lemma in_get:
+  "(r, s') \<in> mres (get s) = (r = s \<and> s' = s)"
+  by (simp add: get_def mres_def)
+
+lemma in_gets:
+  "(r, s') \<in> mres (gets f s) = (r = f s \<and> s' = s)"
+  by (simp add: simpler_gets_def mres_def)
+
+lemma in_put:
+  "(r, s') \<in> mres (put x s) = (s' = x \<and> r = ())"
+  by (simp add: put_def mres_def)
+
+lemma in_when:
+  "(v, s') \<in> mres (when P f s) = ((P \<longrightarrow> (v, s') \<in> mres (f s)) \<and> (\<not>P \<longrightarrow> v = () \<and> s' = s))"
+  by (simp add: when_def in_return)
+
+lemma in_modify:
+  "(v, s') \<in> mres (modify f s) = (s'=f s \<and> v = ())"
+  by (auto simp add: modify_def bind_def get_def put_def mres_def)
+
+lemma gets_the_in_monad:
+  "((v, s') \<in> mres (gets_the f s)) = (s' = s \<and> f s = Some v)"
+  by (auto simp: gets_the_def in_bind in_gets in_assert_opt split: option.split)
+
+lemma in_alternative:
+  "(r,s') \<in> mres ((f \<sqinter> g) s) = ((r,s') \<in> mres (f s) \<or> (r,s') \<in> mres (g s))"
+  by (auto simp add: alternative_def mres_def)
+
+lemma in_liftM:
+  "((r, s') \<in> mres (liftM t f s)) = (\<exists>r'. (r', s') \<in> mres (f s) \<and> r = t r')"
+  by (simp add: liftM_def in_return in_bind)
+
+lemma in_bindE:
+  "(rv, s') \<in> mres ((f >>=E (\<lambda>rv'. g rv')) s) =
+   ((\<exists>ex. rv = Inl ex \<and> (Inl ex, s') \<in> mres (f s)) \<or>
+    (\<exists>rv' s''. (rv, s') \<in> mres (g rv' s'') \<and> (Inr rv', s'') \<in> mres (f s)))"
+  apply (clarsimp simp: bindE_def in_bind lift_def in_throwError)
+  apply (safe del: disjCI; strengthen subst[where P="\<lambda>x. x \<in> mres (f s)", mk_strg I _ E];
+         auto simp: in_throwError split: sum.splits)
+  done
+
+lemmas in_monad = inl_whenE in_whenE in_liftE in_bind in_bindE_L
+                  in_bindE_R in_returnOk in_throwError in_fail
+                  in_assertE in_assert in_return in_assert_opt
+                  in_get in_gets in_put in_when (* unlessE_whenE *)
+                  (* unless_when *) in_modify gets_the_in_monad
```

Comment:
any idea why the `unless_*` are commented out?

Reply:
This comment in Nondet Monad gives a hint
```
(* FIXME lib: remove unlessE_whenE + unless_when here and replace with in_unless lemmas *)
```
It looks like someone removed them for the Trace Monad without adding in the new ones. I might have a quick look to see if they're easy to add.

Reply:
Good news, they were easy to add. You'll see them in the next PR in this chain, along with the equivalent for the nondet monad.